### PR TITLE
Adding read article feature, adding a new view, new template, new route

### LINKED ISSUE #82: Read/View Article feature

### PR DESCRIPTION
**Pull request by Anas-mhameed**
#82 
Implementing read article feature 
Adding new View , new Template , new Route

View and Template Edit required when Authentication implemented.

### CODE REVIEW

**Review by rawadabu — APPROVED** [review]

File: projboard/tests/test_client.py [hunk 1]
```diff
@@ -111,3 +111,7 @@ def test_aboutpage(client):
     template_names = set(tmpl.origin.template_name for tmpl in response.templates)
     # And check if that 'about/about.html' is in the set.
     assert 'about/about.html' in template_names
+
+
+# def test_read_article_page(client):
+#     response = client.get("/article/")
```
Comment (finziamit) [inline projboard/tests/test_client.py]: Is this section will be updated in the future or this comment can be deleted?
Also, isn't it missing `response.status_code 200` assertion?

Comment (Anas-mhameed) [reply projboard/tests/test_client.py]: definitely , but it still draft .

Comment (Anas-mhameed) [reply projboard/tests/test_client.py]: Done

**Review by rawadabu — APPROVED** [review]
LGTM
@Anas-mhameed Just please link it to in issue

File: projboard/views.py [hunk 2]
```diff
@@ -57,3 +58,27 @@ def home_page(request):
...
+    user = User.get_user_by_nickname("User2")
+    article = Article.objects.get(pk=article_pk)
+
+    if request.method == 'POST':
+        try:
+            Like.objects.get(user_id=user.id, article_id=article_pk)
+            Like.delete_like(user_id=user.id, article_id=article_pk)
+
+        except ObjectDoesNotExist:
+            Like.create_like(user_id=user, article_id=article)
```
Comment (finziamit) [inline projboard/views.py]: Can you please explain? does the method check here check the like button click?
If not, wouldn't every entry to the article add\ delete the like?

**Review by finziamit — COMMENTED** [review]
Great work!

**Review by yaelnoyman1 — APPROVED** [review]
great job!
LGTM

File: projboard/views.py (hunk 2 above)
Comment (Anas-mhameed) [reply projboard/views.py]: when client hit like if there is already a like i delete if not i add it .

Comment (Anas-mhameed) [reply projboard/views.py]: yes check like button click

**Review by finziamit — APPROVED** [review]
Great work!

File: projboard/tests/test_client.py [hunk 3]
```diff
@@ -111,3 +111,27 @@ def test_aboutpage(client):
     template_names = set(tmpl.origin.template_name for tmpl in response.templates)
     # And check if that 'about/about.html' is in the set.
     assert 'about/about.html' in template_names
+
+
+@pytest.mark.django_db
+def test_get_readArticlePage(client, article):
```
Comment (Yarboa) [inline projboard/tests/test_client.py]: Please use python naming conventions
https://peps.python.org/pep-0008/#method-names-and-instance-variables

File: projboard/tests/test_client.py [hunk 4]
```diff
@@ -111,3 +111,27 @@ def test_aboutpage(client):
...
+    response = client.get(route)
+    assert response.status_code == 200
+    # Create a set of template names from the templates used in the response
+    template_names = set(tmpl.origin.template_name for tmpl in response.templates)
+    # And check if that 'about/about.html' is in the set.
+    assert 'show/read_article.html' in template_names
+
+
+@pytest.mark.django_db
+def test_post_readArticlePage(client, article):
```
Comment (Yarboa) [inline projboard/tests/test_client.py]: same here

File: projboard/tests/test_client.py (hunk 3 above)
Comment (Yarboa) [inline projboard/tests/test_client.py]: If you create the article, 
and search return home page with article, please assert the fixture content and the result returned by view

File: projboard/tests/test_client.py (hunk 4 above)
Comment (Yarboa) [reply projboard/tests/test_client.py]: This tests created get request, not post, 
It seems as duplicate test with different name

**Review by Yarboa — CHANGES_REQUESTED** [review]

File: projboard/tests/test_client.py (hunk 3 above)
Comment (Anas-mhameed) [reply projboard/tests/test_client.py]: done

File: projboard/tests/test_client.py (hunk 4 above)
Comment (Anas-mhameed) [reply projboard/tests/test_client.py]: done

File: projboard/tests/test_client.py (hunk 3 above)
Comment (Anas-mhameed) [reply projboard/tests/test_client.py]: done

**Review by Shalevro2 — APPROVED** [review]
LGTM

File: projboard/views.py [hunk 5]
```diff
@@ -59,3 +60,27 @@ def home_page(request):
 
 def about_page(request):
     return render(request, 'about/about.html', {})
+
+
+def show_article(request, article_pk):
+
+    # TODO when Authintiaction is implemented => del this line and check wich user is loged in
+    user = User.get_user_by_nickname("User2")
```
Comment (Yarboa) [inline projboard/views.py]: Why do you use single user here? 
Please parse string user nickname in request and read it by view

File: projboard/views.py [hunk 6]
```diff
@@ -59,3 +60,27 @@ def home_page(request):
...
+    article = Article.objects.get(pk=article_pk)
+
+    if request.method == 'POST':
+        try:
+            Like.objects.get(user_id=user.id, article_id=article_pk)
+            Like.delete_like(user_id=user.id, article_id=article_pk)
```
Comment (Yarboa) [inline projboard/views.py]: Why do you delete like for the article on view method?

**Review by Yarboa — CHANGES_REQUESTED** [review]

File: projboard/views.py (hunk 6 above)
Comment (Anas-mhameed) [reply projboard/views.py]: if user hit like and he has been liked the blog before => that means he wants to undo the like .
else he want to like the blog

File: projboard/views.py (hunk 5 above)
Comment (Anas-mhameed) [reply projboard/views.py]: and if i want also to use it in GET method ???

Comment (Yarboa) [reply projboard/views.py]: path('article/<str:article_pk>', views.show_article, name='show_article'),

Maybe only for the sake of test, no need to test likes here at all
test name `show_article`

Just make sure that the relevant article returned with the articl_pk
No need for that like or dislikes

File: projboard/views.py (hunk 6 above)
Comment (Yarboa) [reply projboard/views.py]: I think I missed something,
If a user would like to select/unselect like it should hit GUI element 
Is'nt it? 
I do not see in this MR nothing in templates related to like, did i miss something?

File: projboard/templates/show/read_article.html [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            <p>{{article.content}}</p>
+        </header>
+        <footer>
+            <hr>
+            <form method="POST">
+            <div class="footer">
+                {% csrf_token %}
+                <button class="material-symbols-outlined" type="submit" >
+                    thumb_up
+                </button>
```
Comment (Anas-mhameed) [inline projboard/templates/show/read_article.html]: @Yarboa , here is the GUI that able user to like/dislike the article , its the same button for both funconality (like & dislike) .

File: projboard/views.py (hunk 6 above)
Comment (Anas-mhameed) [reply projboard/views.py]: i have make a commet down, please see it

File: projboard/views.py (hunk 5 above)
Comment (Anas-mhameed) [reply projboard/views.py]: done

File: projboard/templates/searchArticle/search_article.html [hunk 8]
```diff
@@ -40,7 +40,7 @@ <h2>{{article.title | truncatechars:30}}</h2>
                             <p>{{article.content | truncatechars:70}}</p>
                             <footer>
                                 <div class="footer">
-                                    <a href="{% url 'show_article' article.id %}"><button>View Blog</button></a>
+                                    <a href="{% url 'show_article' article_pk=article.id user_nickname='User2'%}"><button>View Blog</button></a>
```
Comment (Yarboa) [inline projboard/templates/searchArticle/search_article.html]: Sorry this is hard coded
Better to give the search option to search the user, and not hard-code it

article/<str:article_pk>/<str:user_nickname>

You could test it also, with different user names,
Part of it return 0 articles and other in DB will return Articles > 0

File: projboard/templates/show/read_article.html (hunk 7 above)
Comment (Yarboa) [reply projboard/templates/show/read_article.html]: Sorry but assume you have few buttons. how do you know which one is submitted? 

               {% csrf_token %}
                <button class="material-symbols-outlined" type="submit">
                    thumb_up
                </button>
                
Are you assuming this is the only POST request parsed to view method?

See Line L71 in the View file, You are not testing like or dislike?
You should have 2 choices for the button, 

This way you can test the post requests for like or dislike an article later on, do you get my point here?

File: projboard/tests/test_client.py [hunk 9]
```diff
@@ -113,3 +114,60 @@ def test_aboutpage(client):
...
+
+
+@pytest.fixture
+@pytest.mark.django_db
+def User2(client):
+    return User.get_user_by_nickname("User2")
+
+
+@pytest.mark.django_db
+def test_read_new_article(client, world_cup_article, User2):
```
Comment (Yarboa) [inline projboard/tests/test_client.py]: The name does not explain what the test need todo test, 
Maybe short description

def test_name()
"""
Test description
"""

File: projboard/tests/test_client.py [hunk 10]
```diff
@@ -113,3 +114,60 @@ def test_aboutpage(client):
...
+    # checking the result of the view return
+    assert response.context['article'] == world_cup_article
+    # getting num of views after viewing the article
+    num_views2 = response.context['article'].num_of_views
+    # checking that num views has been increased
+    assert num_views2 == num_views1 + 1
+
+
+@pytest.mark.django_db
+def test_read_old_article(client, math_article, User2):
```
Comment (Yarboa) [inline projboard/tests/test_client.py]: see above comment

File: projboard/templates/searchArticle/search_article.html (hunk 8 above)
Comment (Anas-mhameed) [reply projboard/templates/searchArticle/search_article.html]: sorry I don't understand 
why do you need me to give it to the search option ?? , i want to parse the user who is authenticated , i don't want the user to choose from list of users ,
i'm parsing the user who did click the button in order to add a new row (if needed) to like Table or delete it if there is already one.

File: projboard/templates/show/read_article.html (hunk 7 above)
Comment (Anas-mhameed) [reply projboard/templates/show/read_article.html]: i'm gonna give an explaination :
in order to make it simple => it' like or unlike button => so i realy don't think there is a need to another button => when user click on it , i check if that user have already liked the blog , if yes so i'm assuming that he want to delete the like  ,else he want to like it => all that done in the view , so i don't need to distingush between like or unlike.

Now , i think you want me to use DELETE Method , is it right ??
im sorry if i didn't get you point

File: projboard/templates/searchArticle/search_article.html (hunk 8 above)
Comment (Yarboa) [reply projboard/templates/searchArticle/search_article.html]: Each that is querying the article will always receive User2 results isnt it? 
It is hardcoded in the template

File: projboard/templates/show/read_article.html (hunk 7 above)
Comment (Yarboa) [reply projboard/templates/show/read_article.html]: No, 
see this tutorial it is a bit different but the same :)
https://www.youtube.com/watch?v=xqFM6ykQEwo

File: projboard/templates/searchArticle/search_article.html (hunk 8 above)
Comment (Anas-mhameed) [reply projboard/templates/searchArticle/search_article.html]: that's Right , are you saying that when someone log in , i should take his data and parse it to the view and then start using it ??

Comment (Yarboa) [reply projboard/templates/searchArticle/search_article.html]: Once the request arrives, to View method it should receive the user

Mostly after login, you are redirected with User logged data to the page
Please refer #87

File: projboard/tests/test_client.py (hunk 10 above)
Comment (Yarboa) [reply projboard/tests/test_client.py]: Still I do not understand the tests here

Comment (Anas-mhameed) [reply projboard/tests/test_client.py]: done

File: projboard/tests/test_client.py (hunk 9 above)
Comment (Anas-mhameed) [reply projboard/tests/test_client.py]: done

File: projboard/templates/searchArticle/search_article.html (hunk 8 above)
Comment (Anas-mhameed) [reply projboard/templates/searchArticle/search_article.html]: done

File: projboard/templates/show/read_article.html (hunk 7 above)
Comment (Anas-mhameed) [reply projboard/templates/show/read_article.html]: done

File: projboard/tests/test_client.py [hunk 11]
```diff
@@ -20,15 +21,35 @@ def test_homepage(client, articles_num):
     assert 'landing/homepage.html' in template_names
 
 
-def test_get_searchpage(client):
-    response = client.get("/search/")
+@pytest.fixture
+@pytest.mark.django_db
+def User2(client):
```
Comment (Yarboa) [inline projboard/tests/test_client.py]: client is not needed here, it is a fixture
You have fixtures already for creating a user
If you need to parse a parameter please fix the fixture 

Please refer this 
https://github.com/redhat-beyond/Smarticle/blob/main/projboard/tests/conftest.py#L11

File: projboard/tests/test_client.py [hunk 12]
```diff
@@ -20,15 +21,35 @@ def test_homepage(client, articles_num):
...
 
 
+@pytest.mark.django_db
+def test_get_empty_searchpage(client):
+    response = client.get("/search/")
+    assert response.status_code == 404
+
+
+@pytest.mark.django_db
+def test_get_ivalid_nickname_searchpage(client, User2):
```
Comment (Yarboa) [inline projboard/tests/test_client.py]: this fixture is not relevant here 
since you are format the string to fakenickname

**Review by Yarboa — COMMENTED** [review]
I do not see test for likes and dislike in this PR

BTW, you are force pushing to the same commit, i can never see the difference
You are squashing only for commits already reviewed.

You should always have at least two commits, it will make the review faster

File: projboard/tests/test_client.py (hunk 12 above)
Comment (Anas-mhameed) [reply projboard/tests/test_client.py]: done

File: projboard/tests/test_client.py (hunk 11 above)
Comment (Anas-mhameed) [reply projboard/tests/test_client.py]: done

**Review by rawadabu — APPROVED** [review]

**Review by Yarboa — APPROVED** [review]
@Anas-mhameed please rebase
LGTM

**Comment by Anas-mhameed** [comment]
@Yarboa done

**Pull request merged**